src/Sarif/Core/Run.cs

```diff
@@ -13,6 +14,109 @@ public partial class Run
+            System.Text.Encoding encoding = null)
+        {
+            if (fileLocation == null) throw new ArgumentNullException(nameof(fileLocation));
+
+            if (this.Files == null || this.Files.Count == 0)
+            {
+                return -1;
+            }
+
+            if (_fileToIndexMap == null)
+            {
+                InitializeFileToIndexMap();
+            }
+
+            // Strictly speaking, some elements that may contribute to a files table 
+            // key are case sensitive, e.g., everything but the schema and protocol of a
+            // web URI. We don't have a proper comparer implementation that can handle 
+            // all cases. For now, we cover the Windows happy path, which assumes that
+            // most URIs in log files are file paths (which are case-insensitive)
+            //
+            // Tracking item for an improved comparer:
+            // https://github.com/Microsoft/sarif-sdk/issues/973
+
+            // When we perform a files table look-up, only the uri and uriBaseId
+            // are relevant; these properties together comprise the unique identity
+            // of the file object. The file index, of course, does not relate to the
+            // file identity. We consciously exclude the properties as well.
+
+            // We will normalize the input fileLocation.Uri to make URIs more consistent
+            // throughout the emitted log.
+            fileLocation.Uri = new Uri(UriHelper.MakeValidUri(fileLocation.Uri.OriginalString));
+
+            var filesTableKey = new FileLocation
+            {
+                Uri = fileLocation.Uri,
+                UriBaseId = fileLocation.UriBaseId
+            };
+
+            int fileIndex;
+            if (!_fileToIndexMap.TryGetValue(filesTableKey, out fileIndex))
+            {
+                if (addToFilesTableIfNotPresent)
+                {
+                    fileIndex = this.Files.Count;
+
+                    string mimeType = Writers.MimeType.DetermineFromFileExtension(filesTableKey.Uri.ToString());
+
+                    var fileData = FileData.Create(
+                        filesTableKey.Uri,
+                        dataToInsert,
+                        mimeType: mimeType,
+                        encoding);
+
+                    fileData.FileLocation = fileLocation;
+
+                    this.Files.Add(new FileData
+                    {
+                        MimeType = mimeType,
+                        FileLocation = filesTableKey
+                    });
+                    _fileToIndexMap[filesTableKey] = fileIndex;
+                }
+                else
+                {
+                    // We did not find the item. The call was not configured to add the entry.
+                    // Return the default value that indicates the item isn't present.
+                    fileIndex = -1;
+                }
+            }
+
+            fileLocation.FileIndex = fileIndex;
```

Comment:
>fileIndex [](start = 37, length = 9)

If you didn't enter the `if` block on Line 62, then this is uninitialized. I'm surprised this compiles.

Reply:
variable declared on line 61. variable is initialized on call on line 62 (where it is used as an 'out' param). the only way line 62 isn't entered is in the event of an exception, in which case line 93 will not execute. now, if this code was executing in a 'finally' clause, you'd see the compilation error.

---
In reply to: [244521899](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244521899) [](ancestors = 244521899)